### PR TITLE
Merge20190712

### PR DESCRIPTION
**Pull request by samtertzakian**
1. Add unit tests for DMF_String.
2.. Clean up support in DMF_String. Make it API work for both Kernel and User-mode.
3. Correct hard coded GUID in DMF_CmApi.
4. Add DMF_File.
5. Add support in DMF_IoctlHandler to allow Client to filter out requests that do not come from Kernel-mode.
6. Update CFU to support transport alignment.

### CODE REVIEW

File: Dmf/Modules.Library/Dmf_File.h [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+// DMF_File_Create()
+//
+DECLARE_DMF_MODULE_NO_CONFIG(File)
+
+// Module Methods
+//
+
+_Must_inspect_result_
+NTSTATUS
+DMF_File_Read(
```
Comment (SurfaceDMFUser1) [inline Dmf/Modules.Library/Dmf_File.h]: We probably want a DMF_File_Write() as well

File: Dmf/Modules.Library/Dmf_String.c [hunk 2]
```diff
@@ -114,49 +114,118 @@ Return Value:
...
+String_MultiStringToWideString(
+    _In_ CHAR* NarrowString
     )
 /*++
 
 Routine Description:
 
-    Perform LEFT case sensitive comparison between string in list and the given string.
-    Comparison is made between the full given string and left of the string in the list.
+    Allocate a buffer for a Wide string and copy into it a converted version of a given
```
Comment (SurfaceDMFUser1) [inline Dmf/Modules.Library/Dmf_String.c]: I would recommend documenting that it is the caller's responsibility to call free() on the string returned.

File: Dmf/Modules.Library/Dmf_String.c [hunk 3]
```diff
@@ -216,8 +287,202 @@ String_WideStringToMultiString(
...
+    if (stringLength + sizeof(WCHAR) > UnicodeString->MaximumLength)
+    {
+        ntStatus = STATUS_BUFFER_TOO_SMALL;
+        TraceEvents(TRACE_LEVEL_ERROR, DMF_TRACE, "UnicodeString's buffer is too small");
+        goto Exit;
+    }
+
+    // Copy the converted string to the destination buffer.
+    //
+    wcscpy_s(UnicodeString->Buffer,
```
Comment (SurfaceDMFUser1) [inline Dmf/Modules.Library/Dmf_String.c]: I believe we are supposed to check the return value (errno_t) here.

File: Dmf/Modules.Library/Dmf_String.c [hunk 4]
```diff
@@ -216,8 +287,202 @@ String_WideStringToMultiString(
...
+    if (stringLength + sizeof(CHAR) > AnsiString->MaximumLength)
+    {
+        ntStatus = STATUS_BUFFER_TOO_SMALL;
+        TraceEvents(TRACE_LEVEL_ERROR, DMF_TRACE, "AnsiString's buffer is too small");
+        goto Exit;
+    }
+
+    // Copy the converted string to the destination buffer.
+    //
+    strncpy_s(AnsiString->Buffer,
```
Comment (SurfaceDMFUser1) [inline Dmf/Modules.Library/Dmf_String.c]: I believe this returns an error code we could check.

File: Dmf/Modules.Library/Dmf_String.h [hunk 5]
```diff
@@ -20,6 +20,40 @@ Module Name:
 
 #pragma once
 
+#if defined(DMF_USER_MODE)
+
+__forceinline
+NTSTATUS
```
Comment (SurfaceDMFUser1) [inline Dmf/Modules.Library/Dmf_String.h]: I would recommend adding _Must_inspect_result_ here

Comment (SurfaceDMFUser1) [reply Dmf/Modules.Library/Dmf_String.h]: Same comment for the other functions returning NTSTATUS codes.

File: Dmf/Modules.Library/Dmf_String.c (hunk 2 above)
Comment (samtertzakian) [reply Dmf/Modules.Library/Dmf_String.c]: Yes, I will do that on next update.

File: Dmf/Modules.Library/Dmf_String.c (hunk 3 above)
Comment (samtertzakian) [reply Dmf/Modules.Library/Dmf_String.c]: I will look into this next update.

File: Dmf/Modules.Library/Dmf_String.c (hunk 4 above)
Comment (samtertzakian) [reply Dmf/Modules.Library/Dmf_String.c]: I will look at this next update.

File: Dmf/Modules.Library/Dmf_String.h (hunk 5 above)
Comment (samtertzakian) [reply Dmf/Modules.Library/Dmf_String.h]: I will do that next update.

**Pull request merged**